src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy

```diff
@@ -237,12 +239,22 @@ class ProtobufPlugin implements Plugin<Project> {
         Configuration compileProtoPath, Collection<Closure> postConfigure) {
       Provider<ProtobufExtract> extractProtosTask =
           setupExtractProtosTask(sourceSet.name, protobufConfig)
-      // In Java projects, the compileClasspath of the 'test' sourceSet includes all the
-      // 'resources' of the output of 'main', in which the source protos are placed.  This is
-      // nicer than the ad-hoc solution that Android has, because it works for any extended
-      // configuration, not just 'testCompile'.
+
+      // Pass include proto files from main to test.
+      // Process resource task contains all source proto files from a proto source set.
+      FileCollection testClassPathConfig = project.objects.fileCollection()
+      if (Utils.isTest(sourceSet.name)) {
+        TaskProvider<ProcessResources> mainProcessResources = project.tasks.named(
+          project.extensions.getByType(SourceSetContainer)
+            .getByName(SourceSet.MAIN_SOURCE_SET_NAME)
+            .processResourcesTaskName,
```

Comment:
Why reference ProcessResources instead of the main proto source set?

Reply:
ProcessResources contains all sources. source code, jars, zips, gradle dependencies, etc. SourceSet contains only sources from source code.

Reply:
Oh, you are also getting the output of extractProtosTask. So the useful parts are coming from how we plumb ProcessResources from the generateProtoTask (which is proto source set + extractProtos):
https://github.com/google/protobuf-gradle-plugin/blob/e20df5b7d70571dc2e7629a77ea1b7e0f5e5862d/src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy#L258-L262

I'm not all that confident that is an equivalent change because the resources will be individual .proto files, not the .jars. I am also not trusting now how people may be modifying gradle tasks.

I can live with this change, but I wonder if it'd be better to just copy from generateProtoTask sourceDirs, the same way as we do for processResources, but into test's generateProtoTask (not extractIncludeProtos). That'd prevent future surprises and seems could be done for Android as well. (We could in the future avoid task configuration for generateProtoTask if we wanted, but I don't know if it really matters.)

Reply:
Two details:

1. Option (1) has more files in it than (2) and (3). (2) and (3) seem equivalent behaviorally at the moment. I can accept any of the answers here. I think (2) and (3) are more bug prone in terms of omission (forget to add something) but I think (1) is more prone to issues in user's builds and most easy to accidentally misuse. For example, with (1) if you add a .proto to src/main/java, you can include that proto in src/test/proto.
2. Shouldn't we add these files as input to generateProtoTask, not to excludeIncludeProtos? This PR is doing excludeIncludeProtos. I'm surprised it works with excludeIncludeProtos, since we're passing .proto files and not .jars. This part is more important to me, because I'm surprised it is working now.

Reply:
> Shouldn't we add these files as input to generateProtoTask, not to excludeIncludeProtos?

What files?

Reply:
Where from jar files can be provided?

Reply:
`mainProcessResources` should be containing the _extracted_ protos and source protos used as input to main's generateProtoTask. It doesn't contain any jars, right? That goes back to the code link above that does `generateProtoTask.get().sourceDirs`.

If it is proto files, it seems we should add those as a dependency to generateProtoTask, not excludeIncludeProtos.

Reply:
It doesnt contains any jars. 
ExcludeIncludeProtos? Do you mean extractIncludeProtos?

Reply:
Main sources should be addes as classpath (include in protoc naming) to test generate task. No reason to add them as sources. It will provide class duplication exception.

Reply:
Yes, I meant extract, not exclude.

If there are no jars, why are these files being passed to setupExtractIncludeProtosTask()? The purpose of that task is to extract proto files from zips. Yes, they shouldn't be a sourceDir for generateProtoTask, but it seems can be an include.

Reply:
Where it pased to extract task? I think i missed something. Could you please explain a bit more.

Reply:
`mainProcessResources` is added to `testClassPathConfig` which is passed:
```groovy
      Provider<ProtobufExtract> extractIncludeProtosTask = setupExtractIncludeProtosTask(
          sourceSet.name, compileProtoPath, testClassPathConfig) // The last argument here
```

And `setupExtractIncludeProtosTask` creates a protobuf extract:
```groovy
      return project.tasks.register(extractIncludeProtosTaskName, ProtobufExtract) {
```

Reply:
https://github.com/google/protobuf-gradle-plugin/blob/master/src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy#L132

If we pass directory, it wont be extracted. Just added to result.

Reply:
We can pass main sources and extractMainProto as includes to generateTestProto and all will works fine. I did this logic in #632